content/2015.09.kinto-encryption-example.rst

```diff
@@ -0,0 +1,207 @@
+      byteArray[i] = str.charCodeAt(i);
+    }
+    return byteArray;
+  };
+  const base64StringToByteArray = (base64) => {
+    return rawStringToByteArray(window.atob(base64));
+  };
+  const byteArrayToBase64String = (buffer) => {
+    var bytes = new Uint8Array(buffer);
+    var binary = '';
+    for (var i=0; i<bytes.byteLength; i++) {
+        binary += String.fromCharCode(bytes[i]);
+    }
+    return window.btoa(binary);
+  };
+
+Next, we need to create a <code>RemoteTransformer</code> that can encode and decode. And for
+that, we need to generate an AES key. We will use that key to encrypt and decrypt
+the data that is passed through the <code>RemoteTransformer</code>:
+
+.. code-block:: javascript
+  // RemoteTransformer:
+  const generateAesKey = () => {
+    // See http://www.w3.org/TR/WebCryptoAPI/#examples-symmetric-encryption
+    return window.crypto.subtle.generateKey({ name: 'AES-CBC', length: 128 },
+         false, ['encrypt', 'decrypt']);
+  };
+
+  const createTransformer = (aesKey) => {
+    const encode = (record) => {
+      const cleartext = rawStringToByteArray(JSON.stringify(record));
+      const IV = window.crypto.getRandomValues(new Uint8Array(16));
+
+      return window.crypto.subtle.encrypt({ name: 'AES-CBC', iv: IV }, aesKey,
+          cleartext).then(ciphertext => {
+        return {
+          id: record.id,
+          ciphertext: byteArrayToBase64String(new Uint8Array(ciphertext)),
+          IV: byteArrayToBase64String(IV)
+        };
+      });
+    };
+
+    const decode = (record) => {
+      const ciphertext = base64StringToByteArray(record.ciphertext);
+      const IV = base64StringToByteArray(record.IV);
+
+      return crypto.subtle.decrypt({ name: 'AES-CBC', iv: IV }, aesKey,
+          ciphertext).then(recordArrayBuffer => {
+
+        return JSON.parse(String.fromCharCode.apply(null,
+            new Uint8Array(recordArrayBuffer)));
+      }, () => {
+        record.undecryptable = true;
+        return record;
+      });
+    };
+
+    return {
+      encode,
+      decode
+    };
+  };
+
+We create two Kinto instances, so that we can test syncing from one instance
+to the other, via the public Kinto demo instance on AWS. Creating multiple Kinto
+instances within the same origin is now possible with the recently added
+<code>dbPrefix</code>
+option:
+
+.. code-block:: javascript
+  // Kinto collection:
+  const createCollection = (transformer, testRun, instanceNo) => {
+    const kinto = new Kinto({
+      dbPrefix: `${testRun}-${instanceNo}`,
+      remote: 'https://kinto.dev.mozaws.net/v1/',
+      headers: {
+        Authorization: 'Basic ' + btoa('public-demo:s3cr3t')
+      }
+    });
+
+    return kinto.collection(`kinto-encryption-example-${testRun}`, {
+      remoteTransformers: [ transformer ]
+    });
+  };
+
+  var coll1, coll2;
+  const prepare = () => {
+    return generateAesKey().then(aesKey => {
+      return createTransformer(aesKey);
+    }).then(transformer => {
```

Comment:
I also like that personally, but I just unlearned that because in Gaia code style the `.then` has to be on the previous line. I'm having trouble getting to a consistent code style with that, because it becomes:
````js
var bla = () => {
  return foo()
    .then(() => {
      bar();
      return baz();
    });
};
````
and you get that weird indentation gap at the bottom. I do agree with you, but I'll stick to the Gaia code style this time. :)